### DISCUSSION TITLE
Using `$wire.$parent.$set()` causes uncaught component bug

### DISCUSSION THREAD

**Discussion by onlyasmalllizard** [livewire/livewire#7907] [1. Help] [answered]
Hello all, I'm working on my first project using Laravel and Livewire 3. I'm having trouble with an uncaught component bug. I've found the line that's causing it, but I don't entirely understand why or the best way to handle it. I was hoping someone here could help me.

I've created a minimal app to recreate the bug. You can see the [entire bug recreation here](https://github.com/onlyasmalllizard/livewire-bug-recreation), but I'll be showing the relevant parts below anyway.

The minimal app has three Livewire components: App, PageOne, and PageTwo. App has a button to change the page, and will either display PageOne or PageTwo as a nested component.

App template:
```
<div>
    <h1>{{ $testProperty }}</h1>

    @if($currentPage === 1)
        <livewire:page-one />
    @else
        <livewire:page-two />
    @endif

    <button wire:click="$set('currentPage', {{ $currentPage === 1 ? 2 : 1 }})">
        {{ $currentPage === 1 ? 'Next' : 'Back' }}
    </button>
</div>
```

App class:
```
<?php

namespace App\Livewire;

use Livewire\Component;

class App extends Component
{
    public int $currentPage = 1;
    public string $testProperty = 'test';
    public function render()
    {
        return view('livewire.app');
    }
}
```

PageOne is basically an empty component to allow us to do the navigation action required to show the bug:
```
<div>
    <h2>Page 1</h2>
</div>

```

In the real app, there are several buttons that are generated using a `@foreach` loop that each open a modal. I'm using the npm package [a11y-dialog](https://a11y-dialog.netlify.app/), which requires me to use a JS script to open and close. I also need to let the parent know which button opened the modal, so I pass an argument into the event to handle the property update at the same time.

The bug occurs without the use of the npm package, so in the minimal recreation, the PageTwo template looks like this:
```
<div>
    <h2>Page 2</h2>

    <button wire:click="$dispatch('testEvent', 'hello!')">Click me!</button>
</div>
@script
<script>
    Livewire.on('testEvent', (message) => {
        $wire.$parent.$set('testProperty', message);
    });
</script>
@endscript

```

If you load the app and navigate to PageTwo, when you click the "Click me!" button, `$testProperty` updates and everything works. The bug occurs if you navigate to PageTwo, navigate back to PageOne, and then back again to PageTwo before clicking the button. When this happens, even though `$testProperty` is still updated, you get the error `Uncaught Could not find Livewire component in DOM tree` in the console. In the real app, this is followed by more buggy behaviour afterwards that seems in line with a missing component key.

It seems to me like using `$wire.$parent.$set()` somehow creates something that Livewire expects to be able to track as a component, but with no way to set a key for it. (Not one that I've been able to find, at least!) If I replace the wire:click action with `$parent.$set('testProperty', 'hello!')`, no errors occur, so it feels like there's some difference between using the `$wire` object and using the magic actions directly in the template that I don't entirely understand.

Does anyone know why this is happening? Is there a way to appease Livewire while using `$wire.$parent.$set()`, or was I going down the wrong path from the start?

**Comment by joshhanley** [comment, marked as answer]
@onlyasmalllizard try putting `wire:key`s on your nested components and see if that fixes the issue. Livewire needs them if you are conditionally adding them. So try something like this:

```blade
@if($currentPage === 1)
        <livewire:page-one wire:key="page-one" />
    @else
        <livewire:page-two wire:key="page-two" />
    @endif
```

If that doesn't work, can you make a https://wirebox.app so we can have a play with it?

**Reply by onlyasmalllizard** [reply]
That's solved the problem, thank you so much for your help! I think it would be worth updating the docs to make it clearer that conditionally added components need a wire:key, even outside of a loop. These were the places in the docs I found while I was trying to trouble shoot:
- [using wire:key in @foreach loops](https://livewire.laravel.com/docs/components#adding-wirekey-to-foreach-loops)
- [using wire:key when rendering children in a loop](https://livewire.laravel.com/docs/nesting#rendering-children-in-a-loop)
I'm not sure if I missed a reference to this though!

**Reply by joshhanley** [reply]
@onlyasmalllizard no worries! Yeah fair, I would argue that the "rendering children in a loop" section covers conditionals too, but that could be made clearer. PRs to update the docs are welcome and appreciated 🙂

**Reply by onlyasmalllizard** [reply]
Sure thing, I'll take a look at it this weekend :)
